### PR TITLE
created/exposed a timeout method for address call

### PR DESCRIPTION
**Pull request by pandey-mohit**
Currently, when we request a resource, address will wait for 30 seconds(default time) to respond. If request is not able to respond within the default time frame, we get ```408 Request Timeout```.
To overcome this, ```timeout``` method has been added which provide a way to alter the time duration.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -304,6 +304,20 @@ address("/price/usd/gbp")
   .post({foo:"bar"})
 ```
 
+### Adding timeout
+When we request for a resource, it waits for a specific time before it sends "408 - request-timeout" error. By adding timeout, we can increase the wait time for the requested resource.
```
Comment (mstade) [inline README.md]: Let's make this read better:

> When requesting a resource  address will by default wait 30 seconds for a response from the resource, before responding with a `408 Request Timeout` error. By calling the `timeout` method on the request, it is possible to increase or decrease this timeout period. The following example shows how to increase the timeout from the default 30 seconds to 60:
> ```javascript
> // increase the timeout to 60000ms (60 sec) for request to resolve
> address("/price/usd/gbp")
>   .timeout(60000)
>   .json()
>   .get()
> ```
>
> The value given to the `timeout` method is in milliseconds.

I'm also not sure whether milliseconds makes sense for this kind of API. The minimum realistic timeout you could set would lie in the ~5-10ms range, and anything sub-second seems rather pointless to be honest. Should we just change this API to accept seconds insteald of milliseconds? That would obviously mean the default would be set to `30` instead of `30000` and the example above would use `60` instead of `60000`.

File: address.js [hunk 2]
```diff
@@ -203,6 +210,7 @@ define(function(require) {
         , body : body
         , context : context
         , origin: origin
+        , timeout: timeout
```
Comment (mstade) [inline address.js]: I don't particularly like how we keep polluting the request object with additional data just to use it as a transport mechanism, it smells and it smells bad. That said, the precedence was set with `context` and `origin`, and doing something different with `timeout` is probably overkill for this PR, but we should consider rewriting this altogether at some point.

File: test/address.test.js [hunk 3]
```diff
@@ -219,6 +220,7 @@ define(function(require) {
             , body : undefined
             , context: document.body
             , origin: undefined
+            , timeout: 30000
```
Comment (mstade) [inline test/address.test.js]: It looks like these are just added gratuitously and then never used, so they should be removed again.

Comment (pandey-mohit) [reply test/address.test.js]: ```timeout``` is passed to middleware using req, so we need to add them, else test case will fail.
 
Also, the test case has been updated for "configured web.req" to validate the timeout method with increased value.

File: address.js (hunk 2 above)
Comment (pandey-mohit) [reply address.js]: Agreed.

File: README.md (hunk 1 above)
Comment (pandey-mohit) [reply README.md]: Existing libraries, like d3, jQuery etc. are using time in milliseconds and also a standard way to provide duration in timeouts.

File: test/address.test.js (hunk 3 above)
Comment (mstade) [reply test/address.test.js]: Ok, I will need to run these locally to check how this works, bear with me.

File: README.md [hunk 4]
```diff
@@ -304,6 +304,19 @@ address("/price/usd/gbp")
   .post({foo:"bar"})
 ```
 
+### Adding timeout
+When requesting a resource address will by default wait 30 seconds for a response from the resource, before responding with a ```408 Request Timeout``` error. By calling the ```timeout``` method on the request, it is possible to increase or decrease this timeout period. The following example shows how to increase the timeout from the default 30 seconds to 60:
+
+```javascript
+// increase the timeout to 60000ms (60 sec) for request to resolve
+address("/price/usd/gbp")
+  .timeout(60000)
```
Comment (mstade) [inline README.md]: I still think we should change this API to accept seconds, rather than milliseconds. As a user, you could still increase granularity by just using fractions (e.g. `0.5` for a half second) if we just multiply the parameter value by `1000`. There's a technical risk of inaccuracies I suppose because we'd be using floats, but it's not like the timeout is always accurate anyway so I don't think it matters.

File: README.md [hunk 5]
```diff
@@ -304,6 +304,19 @@ address("/price/usd/gbp")
   .post({foo:"bar"})
 ```
 
+### Adding timeout
+When requesting a resource address will by default wait 30 seconds for a response from the resource, before responding with a ```408 Request Timeout``` error. By calling the ```timeout``` method on the request, it is possible to increase or decrease this timeout period. The following example shows how to increase the timeout from the default 30 seconds to 60:
```
Comment (mstade) [inline README.md]: We should change the text here to say `wait at least` rather than just `wait`, since the timeout detection isn't always accurate – if there's something holding up the main thread for instance we may well be waiting longer. It's kind of an academic distinction, but still.

Comment (mstade) [reply README.md]: Let's also change `before responding with` to `before returning`, it reads better.

Comment (mstade) [reply README.md]: Also there's no need to use triple backticks when inlining quotes, single backticks will do fine.

File: README.md (hunk 4 above)
Comment (pandey-mohit) [reply README.md]: Updated `timeout` method to accept seconds instead of milliseconds.

File: middleware.js [hunk 6]
```diff
@@ -7,12 +7,12 @@ define(function(require) {
   return {
 
     requestTimeout : function(req, res, next) {
-
       var responded
+        , interval = req.timeout * 1000
```
Comment (mstade) [inline middleware.js]: I know there’s some validation of this value in address.js, which is all well and good, but we should take care to validate these values here as well where they are actually used. If req.timeout isn’t a number, this will result in NaN and potentially difficult to debug behavior. Likewise, negative values can have confusing effects. We should take care here that the value must be a positive number, in the range `0-Infinity`, and if indeed it is not then and only then do we fall back to a standard timeout. That also implies that the standard value of 30sec shouldn’t be defined in address.js, but in this middleware and potentially exposed for the benefit of reuse by address.js.

File: address.js [hunk 7]
```diff
@@ -31,6 +31,7 @@ define(function(require) {
         , callback
         , target
         , into
+        , timeout = 30
```
Comment (mstade) [inline address.js]: See my comment below about exporting this value as a constant from the middleware instead.

File: middleware.js (hunk 6 above)
Comment (mstade) [reply middleware.js]: Also, we should add some tests to cover these inputs, particularly the cases of negative values an non-number values, including `null` and `undefined`.

Comment (pandey-mohit) [reply middleware.js]: Added timeout validation and test cases.

File: address.js [hunk 8]
```diff
@@ -191,6 +194,11 @@ define(function(require) {
         api.method('remove').body(body)()
       }
 
+      api.timeout = function(v) {
+        timeout = util.isValidNumber(v) ? v : timeout
```
Comment (mstade) [inline address.js]: No no, the validation needs to happen in the middleware, where it's *used*. The point of validating where the value is being used is because even if it's validated here it can still be changed in transit. The validation needs to happen in this function, before the value is used: https://github.com/pandey-mohit/address/blob/9c92f228609002af6e9fd32df940463ff47311ea/middleware.js#L11

File: util.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+define(
+  []
+, function() {
+
+    function isValidNumber(num) {
+      return (/^[+]?([0-9]+(?:[\.][0-9]*)?|\.[0-9]+)$/).test(num)
```
Comment (mstade) [inline util.js]: Oh no, don't use regex for this, it's not a very good fit for the job. Here's a more solid function:

```javascript
function isValidTimeout(n) {
  return !isNaN(parseFloat(n)) && isFinite(n) && Math.max(0, n) === n
}
```

The first two clauses will check that `n` is a number, and the last will check that it's in the range of `0` and all finite numbers above.

Also, please don't make this a separate module, just put it in the middleware module, where it's in scope of the timeout middleware. There's no reason to make this a separate module when it's not reused anywhere else.

File: test/address.test.js [hunk 10]
```diff
@@ -198,13 +199,97 @@ define(function(require) {
...
+          .on('done', cb)
+
+        update()
+
+        web.req.should.have.been.calledOnce
+        web.req.should.have.been.calledWith(req)
+        cb.should.have.been.calledOnce
+      })
+
+      it('should call configured request and callback with the default timeout if negative value is provided', function() {
```
Comment (mstade) [inline test/address.test.js]: No, we should throw an error if the timeout is invalid.

File: test/address.test.js [hunk 11]
```diff
@@ -198,13 +199,97 @@ define(function(require) {
...
+          .on('done', cb)
+
+        update()
+
+        web.req.should.have.been.calledOnce
+        web.req.should.have.been.calledWith(req)
+        cb.should.have.been.calledOnce
+      })
+
+      it('should call configured request and callback with the default timeout if null value is provided', function() {
```
Comment (mstade) [inline test/address.test.js]: No, again, we should throw an error. Invalid values shouldn't transform to some other value.

File: util.js (hunk 9 above)
Comment (pandey-mohit) [reply util.js]: Libraries, like `jQuery. d3` etc. doesn't force user and internally check whether the provided timeout is correct or not. If correct, then it will add the timeout. Else, it will set the default value.

Ref: 
[jquery-timeout!](https://github.com/jquery/jquery/blob/master/src/ajax.js#L687)
[d3-timeout!](https://github.com/d3/d3-request/blob/master/src/request.js#L112)

If agreed, then we can update the current code(as per d3, jQuery, like, `req.timeout > 0` and add the same check in middleware.js, instead of adding `isValidTimeout` method, so that if correct timeout is passed, we will use it. Else, will use our default value of `30 milliseconds`.

If not, I am happy to update codebase as per above requested changes. Also, please let me know the error which we need to send in case of invalid timeout. Can we use `"400": ["bad-request", "client-error"]`.

Thoughts??

File: address.js (hunk 8 above)
Comment (pandey-mohit) [reply address.js]: IMHO, the validation needs to be happen at the first place. Also, I am not sure how it can be changed in transit.
But, I am happy to make changes and add it to middleware, since asked.

Comment (mstade) [reply address.js]: You could add it to both, but it’d be overkill. Since the timeout is exposed on the request object it can be changed at any point during the request handling process, which of course includes the end request handler but it also includes all middleware, and whatever meddling that address does itself. It’d be trivial to write a middleware function that messes up the value, which is why validation at use is the only place it truly matters.

Throwing an error at that point however may be less convenient, since it very likely wouldn’t be in the same stack as where the value was originally set on the request object, which incidentally is an argument for doing the validation twice.

File: util.js (hunk 9 above)
Comment (mstade) [reply util.js]: Ahh yes, that’s an elegant check, let’s do that – nice find!

As for the error: just throw, don’t send an error response.

Comment (pandey-mohit) [reply util.js]: Updated the codebase as requested.

I have also added an error response `"400"` with `Invalid timeout` message because of the following reasons:
1. If we send `res(null)` without any error response, then code will break at different levels in address lib's like, in `logger` method and `is-view.js` etc..., where we are not handling an empty response.
2. By sending with `"400"`, we let user to know that its a `bad request` and message will tell it because of timeout.

Please let me know in case it needs further updation.

File: test/address.test.js [hunk 12]
```diff
@@ -198,16 +199,54 @@ define(function(require) {
...
       })
 
+      it('should return error when negative timeout value is provided', function() {
+        var cb = sinon.spy()
+          , responseBody = {
+              message: 'Invalid timeout: must be an integer value greater than 0'
+            }
+
+        response.statusCode = 400
+        response.body = responseBody
```
Comment (mstade) [inline test/address.test.js]: I didn't quite understand what this was so I had to do a double take and make sure I understood how the tests actually function, and as it turns out this right here means we're not actually testing this functionality. This test, and the other one that tests for null timeout values, isn't actually testing to make sure the middleware blows up during the request – really all we're doing here is pretend making a request and returning an exact response.

This seems indicative of a larger problem with this test suite to be honest, where we're just testing the API but not actually the functionality. The reason I found this out is because I didn't expect there to be response, I figured that when the request is actually made an exception would be thrown. This is a jarring experience, but frankly, an invalid timeout value should be a hard error – not a response – such that it can be caught in development rather than production.

Here's what I recommend: move these tests that check the middleware into its own file, called middleware.test.js, and set it up like so:

```javascript
define(function(require) {
  var sinon = require('sinon')
    , nap = require('@websdk/nap')
    , address = require('address')
    , middleware = require('middleware')

  describe('Middleware', function() {
    it('should throw error when negative timeout value is provided', function() {
      var resource = sinon.spy()
      var cb = sinon.spy()
      var error

      try {
        address('/wibble/{id}')
          .web(
            nap.web()
              .use(middleware.requestTimeout)
              .resource('wibble', '/wibble/{id}', resource)
          )
          .param('id', '123')
          .timeout(-10)
          .on('done', cb)()
      } catch (e) {
        error = e
      } finally {
        cb.should.not.have.been.called
        resource.should.not.have.been.called
        error.message.should.equal('Invalid timeout: must be an integer value greater than 0')
      }
    })

    it('should return error when null timeout value is provided', function() {
      var resource = sinon.spy()
      var cb = sinon.spy()
      var error

      try {
        address('/wibble/{id}')
          .web(
            nap.web()
              .use(middleware.requestTimeout)
              .resource('wibble', '/wibble/{id}', resource)
          )
          .param('id', '123')
          .timeout(null)
          .on('done', cb)()
      } catch (e) {
        error = e
      } finally {
        cb.should.not.have.been.called
        resource.should.not.have.been.called
        error.message.should.equal('Invalid timeout: must be an integer value greater than 0')
      }
    })

    it('should set timeout to 30 seconds by default', function() {
      address('/wibble/{id}')
          .web(
            nap.web()
              .use(middleware.requestTimeout)
              .resource('wibble', '/wibble/{id}', assertDefaultTimeout)
          )
        .param('id', '123')
        .on('done', function() {})()

      function assertDefaultTimeout(req, res) {
        req.timeout.should.equal(30)
      }
    })
  })
})
```

This will test the middleware proper, since this is more of an integration test than an API test, like those in address.test.js. Of course, this also means removing the middleware tests from address.test.js.

**Comment by mstade** [comment]
So it turns out I was able to push the changes to your branch @pandey-mohit, not sure how but hey it worked. Maybe it's because I'm a maintainer of this project? Anyway, I pushed the changes to the tests and this all seems to work and look just fine now.

Is there anything else you'd like to add @pandey-mohit? Otherwise I'll bump and publish a new minor version as soon as you give the go ahead.

**Comment by pandey-mohit** [comment]
Thanks @mstade.

I have also verified the changes at my end and everything seems to be connecting well. 

Also, i found one issue which is related to existing timeout functionality. Since default timeout is of 30 sec and if request is not served within the provided time, then we get `408 - [request-timeout]` error, but request doesn't get terminated and still being served in background. 
Not sure, whether its the desired functionality or a gap, but we should look into it, maybe in a separate PR. 

I think we are good to go and create a new version. Please publish a new version so that we can able to integrate those changes.

Thanks again for reviewing :)
